libs/go/libp2p_node/dht/dhtpeer/dhtpeer.go

```diff
@@ -1433,72 +1434,67 @@ func (dhtPeer *DHTPeer) handleAeaNotifStream(stream network.Stream) {
-				lerror(nil).
-					Msgf("timeout: notifying peer %s haven't been added to DHT routing table",
-						stream.Conn().RemotePeer().Pretty())
-				return
-			case <-time.After(time.Millisecond * 5):
-			}
+	opLatencyRegister, _ := dhtPeer.monitor.GetHistogram(metricOpLatencyRegister)
+	timer := dhtPeer.monitor.Timer()
+	start := timer.NewTimer()
+
+	// workaround: to avoid getting `failed to find any peer in table`
+	//  when calling dht.Provide (happens occasionally)
+	ldebug().Msg("waiting for notifying peer to be added to dht routing table...")
+	ctx, cancel := context.WithTimeout(
+		context.Background(),
+		routingTableConnectionUpdateTimeout,
+	)
+	defer cancel()
+	for dhtPeer.dht.RoutingTable().Find(stream.Conn().RemotePeer()) == "" {
+		select {
+		case <-ctx.Done():
+			lerror(nil).
+				Msgf("timeout: notifying peer %s haven't been added to DHT routing table",
+					stream.Conn().RemotePeer().Pretty())
+			return
+		case <-time.After(time.Millisecond * 5):
 		}
+	}
 
-		if dhtPeer.myAgentAddress != "" {
-			err := dhtPeer.registerAgentAddress(dhtPeer.myAgentAddress)
+	if dhtPeer.myAgentAddress != "" && !dhtPeer.IsAddressAnnounced(dhtPeer.myAgentAddress) {
+		err := dhtPeer.registerAgentAddress(dhtPeer.myAgentAddress)
+		if err != nil {
+			lerror(err).Str("op", "notif").
+				Str("addr", dhtPeer.myAgentAddress).
+				Msgf("while announcing my agent address")
+			return
+		}
+	}
+	if dhtPeer.enableRelay {
+		dhtPeer.dhtAddressesLock.RLock()
+		for addr := range dhtPeer.dhtAddresses {
+			err := dhtPeer.registerAgentAddress(addr)
 			if err != nil {
 				lerror(err).Str("op", "notif").
-					Str("addr", dhtPeer.myAgentAddress).
-					Msgf("while announcing my agent address")
-				return
+					Str("addr", addr).
+					Msg("while announcing relay client address")
 			}
 		}
-		if dhtPeer.enableRelay {
-			dhtPeer.dhtAddressesLock.RLock()
-			for addr := range dhtPeer.dhtAddresses {
-				err := dhtPeer.registerAgentAddress(addr)
-				if err != nil {
-					lerror(err).Str("op", "notif").
-						Str("addr", addr).
-						Msg("while announcing relay client address")
-				}
+		dhtPeer.dhtAddressesLock.RUnlock()
+	}
+	if dhtPeer.delegatePort != 0 {
+		dhtPeer.tcpAddressesLock.RLock()
+		for addr := range dhtPeer.tcpAddresses {
+			err := dhtPeer.registerAgentAddress(addr)
+			if err != nil {
+				lerror(err).Str("op", "notif").
+					Str("addr", addr).
+					Msg("while announcing delegate client address")
 			}
-			dhtPeer.dhtAddressesLock.RUnlock()
 		}
-		if dhtPeer.delegatePort != 0 {
-			dhtPeer.tcpAddressesLock.RLock()
-			for addr := range dhtPeer.tcpAddresses {
-				err := dhtPeer.registerAgentAddress(addr)
-				if err != nil {
-					lerror(err).Str("op", "notif").
-						Str("addr", addr).
-						Msg("while announcing delegate client address")
-				}
-			}
-			dhtPeer.tcpAddressesLock.RUnlock()
+		dhtPeer.tcpAddressesLock.RUnlock()
 
-		}
-		duration := timer.GetTimer(start)
-		opLatencyRegister.Observe(float64(duration.Microseconds()))
```

Comment:
should we not keep these timers?

Reply:
it's saved, just indent changes so github shows as removed